File: compiler/bench/fibmemo.fs
\ expected: -2872092127636481573
\ Fibonacci with memoization, fib(10000) mod 2^63 (wraps)

10001 constant MEMO-SIZE
create memo MEMO-SIZE cells allot

: fibmemo-init ( -- )
  MEMO-SIZE 0 do 0 memo i cells + ! loop
  1 memo 1 cells + ! ;

: fibmemo ( n -- f )
  dup 2 < if exit then
  2 begin
    dup 3 pick <= while
    dup 1- cells memo + @
    over 2 - cells memo + @
    + over cells memo + !
    1+
  repeat drop
  cells memo + @ ;

: main
  fibmemo-init
  10000 fibmemo . cr ;

File: compiler/bench/fibmemo.c
// expected: -2872092127636481573
#include <stdio.h>
#define MEMO_SIZE 10001
long memo[MEMO_SIZE];

void fibmemo_init() {
    for (int i = 0; i < MEMO_SIZE; i++) memo[i] = 0;
    memo[1] = 1;
}

long fibmemo(long n) {
    if (n < 2) return n;
    for (long i = 2; i <= n; i++) {
        memo[i] = memo[i-1] + memo[i-2];
    }
    return memo[n];
}

int main() {
    fibmemo_init();
    printf("%ld\n", fibmemo(10000));
    return 0;
}
